src/process/io/s3/S3Source.h

```diff
@@ -0,0 +1,222 @@
+/**
+ * Copyright (c) 2019 DKFZ - ODCF
+ *
+ * Distributed under the MIT License (license terms are at https://github.com/dkfz-odcf/FastqIndEx/blob/master/LICENSE.txt).
+ */
+
+#ifndef FASTQINDEX_S3Source_H
+#define FASTQINDEX_S3Source_H
+
+#include "S3Config.h"
+#include "S3GetObjectProcessWrapper.h"
+#include "process/io/Source.h"
+#include "process/io/StreamSource.h"
+#include "FQIS3Client.h"
+
+#include <aws/core/Aws.h>
+#include <aws/core/auth/AWSCredentialsProvider.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/GetObjectRequest.h>
+#include <fstream>
+#include <iostream>
+#include <semaphore.h>
+
+/**
+ * This class enables you to get data from an S3 bucket. The way, how we retrieve data, might seem odd. Unfortunately,
+ * we experienced quite a number of issues with the Amazon AWS SDK and eventually decided to do everything like this.
+ * - The SDK does not allow to work directly with the data stream. Therefore, we decided to download the data to
+ *   a named pipe in the tmp folder and access this with a StreamSource
+ * - Due to this, it is not possible to seek within an S3 file stream! Fortunately, the StreamSource has some built-in
+ *   buffer to deal with this.
+ * - It is not possible to abort a running request! What we did first was download with GetObjectAsync. However this
+ *   lead to errors, when we closed the pipe or tried to close the stream. Worst case: The application crashed, best
+ *   case: It didn't crash but always showed an error (SIGPIPE / SIGABRT). Personally, my trust in application is not
+ *   very high, if it shows me such types of errors.
+ *   Eventually we moved the download to a separate process. It works but has no real error handling by now. But: It
+ *   swallows any bad error codes and messages like a broken pipe (which is not nice but intentional in this case).
+ * Unfortunately, we are not able to really know, how much data we need for extraction and we need the possibility to
+ * abort a Get request at any time.
+ *
+ * With the current implementation, things work so far but we still need to think about error handling and reporting.
+ */
+class S3Source : public Source {
+
+private:
+
+    bool _isOpen{false};
+
+    bool sizeRequested{false};
+
+    int64_t _size{0};
+
+    int64_t position{0};
+
+    FQIS3Client fqiS3Client;
+
+    path fifo;
+
+    shared_ptr<S3GetObjectProcessWrapper> s3GetObjectWrapper;
+
+    FStream *s3FStream{nullptr};
+
+    ifstream stream;
+
+    shared_ptr<StreamSource> streamSource;
+
+    mutex signalDisablingMutex;
+
+public:
+
+    const path &getFifo() const;
+
+    const ifstream &getStream() {
+        return stream;
+    }
+
+    static shared_ptr<S3Source> from(const string &file, const S3ServiceOptions &s3ServiceOptions) {
+        return make_shared<S3Source>(file, s3ServiceOptions);
+    }
+
+    /**
+     * Create an instance of this object with a path source. This will be read by fread and so on.
```

Comment:
The comment is unnecessary. It is a constructor. Information about interna should be internal with the implementation code, not the method description.